src/network/networkd-dhcp-static-lease.c

```diff
@@ -0,0 +1,269 @@
+#include "networkd-util.h"
+#include "networkd-dhcp-static-lease.h"
+#include "memory-util.h"
+#include "strv.h"
+#include "utf8.h"
+
+DHCPStaticLease *dhcp_static_lease_free(DHCPStaticLease *static_lease) {
+        if (!static_lease)
+                return NULL;
+
+        if (static_lease->network) {
+                LIST_REMOVE(dhcp_static_leases, static_lease->network->dhcp_static_leases, static_lease);
+                assert(static_lease->network->n_dhcp_static_leases > 0);
+                static_lease->network->n_dhcp_static_leases--;
+
+                if (static_lease->section)
+                        hashmap_remove(static_lease->network->dhcp_static_leases_by_section, static_lease->section);
+        }
+
+
+
+        network_config_section_free(static_lease->section);
+        free(static_lease->client_id.data);
+        return mfree(static_lease);
+}
+
+static int dhcp_static_lease_fill(
+                DHCPClientId *client_id,
+                const struct in_addr *address,
+                DHCPStaticLease **ret) {
+        DHCPStaticLease *p;
+
+        assert_return(ret, -EINVAL);
+        assert_return(address, -EINVAL);
+        assert_return(client_id, -EINVAL);
+
+        p = new(DHCPStaticLease, 1);
+        if (!p)
+                return -ENOMEM;
+
+        *p = (DHCPStaticLease) {
+                .client_id = *client_id,
+                .address = le32toh(address->s_addr)
+        };
+
+        *ret = TAKE_PTR(p);
+        return 0;
+}
+
+int dhcp_static_lease_new(DHCPStaticLease **ret) {
+        struct in_addr address = {};
+        DHCPClientId client_id = {};
+
+        assert_return(ret, -EINVAL);
+
+        return dhcp_static_lease_fill(&client_id, &address, ret);
+}
+
+int dhcp_static_lease_set_client_id_by_mac(
+                DHCPStaticLease *lease,
+                const uint8_t *mac_addr,
+                size_t mac_addr_size) {
+
+        uint8_t *data;
+
+        assert_return(lease, -EINVAL);
+        assert_return(mac_addr, -EINVAL);
+        assert_return(mac_addr_size, -EINVAL);
+        
+        data = new(uint8_t, mac_addr_size + 1);
+        if (!data)
+                return -ENOMEM;
+
+        /* set client id type to 1: Ethernet Link-Layer (RFC 2132) */
+        data[0] = 0x01;
+        memcpy(data + 1, mac_addr, mac_addr_size);
+        
+        free_and_replace(lease->client_id.data, data);
+
+        lease->client_id.length = mac_addr_size + 1;
+
+        return 0;
+}
+
+int dhcp_static_lease_set_address(DHCPStaticLease *lease, const struct in_addr *address) {
+        assert_return(address, -EINVAL);
+
+        lease->address = address->s_addr;
+
+        return 0;
+}
+
+int dhcp_static_lease_unset_client_id(DHCPStaticLease *l) {
+        assert_return(l, -EINVAL);
+
+        l->client_id.data = mfree(l->client_id.data);
+        l->client_id.length = 0;
+
+        return 0;
+}
+
+int dhcp_static_lease_section_verify(DHCPStaticLease *l) {
+        assert_return(l, -EINVAL);
```

Comment:
Do not return negative errno but just abort the execution. That is, please use `assert()` here. Otherwise we call `free()` for `NULL` in `network_verify()`.